### ISSUE TITLE
Nested workflow too many open files (solved with limitation)

### ISSUE THREAD

**Issue by gaow** [vatlab/sos#1376]
Separated from #1374 

here is the complete log:

[hybrid.log.gz](https://github.com/vatlab/sos/files/4702403/hybrid.log.gz)

and the last few lines are the errors:

```
DEBUG: EXECUTOR - Stop controller from 10339
DEBUG: EXECUTOR - disconntecting master
DEBUG: CONTROLLER - 10339 : closes socket with handler 17 (344 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 20 (343 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 23 (342 left)
DEBUG: CONTROLLER - controller stopped 10339
DEBUG: CONTROLLER - 10339 : closes socket with handler 39 (341 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 32 (340 left)
DEBUG: CONTROLLER - Disconnecting sockets from 10339
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/__main__.py", line 642, in cmd_run
    executor.run(args.__targets__, mode=config['run_mode'])
Traceback (most recent call last):
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/__main__.py", line 642, in cmd_run
    executor.run(args.__targets__, mode=config['run_mode'])
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/workflow_executor.py", line 347, in run
    return self.run_as_master(targets=targets, mode=mode)
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/workflow_executor.py", line 1649, in run_as_master
    raise exec_error
sos.executor_utils.ExecuteError: [hybrid]: Too many open files
[hybrid]: Exits with 1 pending step (hybrid_4)
ERROR: [hybrid]: Too many open files
[hybrid]: Exits with 1 pending step (hybrid_4)
```

**Comment by gaow** [comment]
Notice this problem can be reliably reproduced in our analysis. Here we have about 500 substeps for the nested workflow which is not a lot. In `stress-test` repo we have examples with many more substeps per nested workflow but it was never an issue. The way we trigger nested workflows are different between the two cases which might be the reason for the problem.

**Comment by BoPeng** [comment]
So you are still using midway2 at UChicago? Can I log in to have a look?

**Comment by BoPeng** [comment]
I can reproduce the problem with the following example

```
[c]
parameter: t=int
print(t)

[a]
parameter: t = int
sos_run('c', t=t)

[default]
input: for_each=dict(i=range(10000))
sos_run('a', t=i)
```

So basically a large number of nested workflows are pending...

**Comment by BoPeng** [comment]
Thinking hard,

now we need 8 ports for 8 pending jobs,

```
master port  1  ---    worker1. pending 1
master port  2  ---    worker1. pending 2
master port  3  ---    worker1. pending 3
master port  4  ---    worker2. pending 1
master port  5  ---    worker2. pending 2
master port  6  ---    worker2. pending 3
master port  7  ---    worker2. pending 4
master port  8  ---    worker3. pending 1
```
Giving that at most one job will be running at a time, we actually only need one socket per worker to do all the necessary communication. The problem is how to identify the pending jobs...

**Comment by BoPeng** [comment]
Started working on it but changing the way workers and scheduler communicate is not easy and I could only get the easy cases run at this point. I will have to work on a COVID-related project for now and will get back to this as soon as i can.

**Comment by gaow** [comment]
@BoPeng Thank you! At this point we are running those analysis without using that nested workflow; then manually run separate analysis for it based on `if ... else` conditions, so the feature wouldn't impact us immediately. But it would be real nice to have it fixed and work for SoS.

**Comment by BoPeng** [comment]
The bunk part has been done but workflows stop here and there. There should be only one or two hiccups but I am asked to expand my [outbreak simulator](https://github.com/ictr/covid19-outbreak-simulator) to simulate the best strategy for community-based PCR testing.

**Comment by BoPeng** [comment]
Just note here that there will be some performance compromise because now whenever the worker starts working on another pending job, it will has to notify the master that the same port will be handling another step, so that the master knows where the result come from. This can double the number of messages sent around although the notification messages are very short.

Nested workflows still do not work due to the change of messaging.

**Comment by BoPeng** [comment]
It was too soon to blame myself for choosing an approach that used too many sockets as it turns out that might be necessary. The problem is not only with sending job requests from worker to master, which we could let master know before hand who is working, it is also the problem for sending request from master to worker.

In the nested workflow case, IF the two workflows are on the same worker, (we can force it by using `-j1`),

```
[c]
parameter: t=int
print(t)

[a]
parameter: t = int
sos_run('c', t=t)

[default]
input: for_each=dict(i=range(1))

sos_run('a', t=i)
```

what will happen is that the master will send result from the nested nested workflow to the worker, while there are two workflows `a` and `c` that are both waiting, and if `a` get the result, `c` will not have a change to do it.

If two workflows both need a socket to communicate, then we will be using too many sockets if there are many pending workflows.

**Comment by gaow** [comment]
I'm not sure if I follow this entirely. You are saying that at least with current design of nested workflow, we either have to indeed open many sockets, or force using smaller numbers of sockets with `-j` and suffer from performance loss? I tried example above with `-j20` and `-j1`. I dont see obvious differences in the outcome.

**Comment by BoPeng** [comment]
I have found the problem. When we do

```
[default]
input: for_each=dict(i=range(1000))
sos_run('a', t=i)
```
sos will allow the subworkflows to be executed in parallel, and in this case it will send all 1000 subworkflows to the controller. Then while a few subworkflows are being executed, in the case of nested nested workflow, more subworkflows enters the controller, and some of them will be dispatched to workers, occupying sockets, when they will not actually be completed after quite a while.

The solution is actually quite simple, do not send 1000 subworkflows to the controller. In a patch that I am testing, at most `-j` subworkflows will be submitted at a time so that even in the case of nested nested workflow, the controller has mostly workflows being executed.

**Comment by BoPeng** [comment]
We did not see the problem before because the controller will submit steps before subworkflow, so if there is no nested nested workflow, the execution of subworkflows will generate a bunch of steps, which will be executed before other subworkflows. In this way not many subworkflows will be sent to workers at the same time.

Only in the case of nested nested workflow do we see an overflow of pending workflows, and too many open files.

I am writing down the details to remind myself if this piece of code needs to be revisited later.

**Comment by BoPeng** [comment]
@gaow Please test and see if this fixes the problem.

**Comment by gaow** [comment]
>  In a patch that I am testing, at most -j subworkflows will be submitted at a time 

Great this makes sense now. I was under the impression that we'll either not limit it, or limit it to 1 ...  

> I am writing down the details to remind myself if this piece of code needs to be revisited later.

Is there something else that needs revisiting, or it is just tiding up the codes? I am submitting job right now. I'll report here if there is still a problem.

**Comment by gaow** [comment]
Cannot really test it at this point due to https://github.com/vatlab/sos/issues/1375#issuecomment-637237868

**Comment by BoPeng** [comment]
Just because I went a long way trying to address the socket problem by changing how the sockets were generated and used, but finally found out the reason for the original implementation. :-(

So it makes sense to write down reasons behind the code..

**Comment by gaow** [comment]
It now roughly works -- the problem is that it does not seem to submit to the cluster enough jobs as I specified. My maximum allowed job is 60 and I have 500 tasks to submit from this step. For other pipelines it is able to submit up to that many in the queue. But now I'm looking at only 9 jobs submitted. The scheduler wont submit new jobs until after these complete.

I can try to come up an MWE tmr. But I'm reporting here just in case this is an obvious fix.

**Comment by gaow** [comment]
Hmm I might see what's going on -- number of tasks submitted to cluster is `-J M` and should be independent from `-j N`. But for nested workflow we are limited by `-j N` to create N processes and each process will submit a task. So the number of tasks is now limited by `-j`?

**Comment by BoPeng** [comment]
Yes, let me have a detailed look tomorrow. It is possible that there is not enough nested workflow to feed to the task.

**Comment by BoPeng** [comment]
If I understand correctly, 

* You have a large number of subworkflows, each submit a task. 
* Your job queue is supposed to accommodate, say, 500 tasks. 
* Your subworkflow will not complete until it can hear back from the tasks.
* At most `-j` subworkflows can be pending (waiting for task results) so at most `-j` tasks can be running.

So the key problem is that 

1. In the case of substeps, substeps are actually completed (no longer occupying port) and the step port is waiting for the task results. So all tasks could be submitted at once.

2. In the case of subworkflows -> step -> task, the step is fine but the number of subworkflows limits the number of tasks. I do not see a good way to get around of this.

**Comment by gaow** [comment]
My workflow looks exactly like this example: https://github.com/vatlab/sos/issues/1376#issuecomment-636276661

But in the steps `a` and `c` I use `task` to submit to the cluster. My queue, configured by `hosts` in SoS configuration file, can allow for a maximum of 100 tasks. These steps are now used as nested workflow (subworkflow) and they will only submit (at most) `-j` substeps at a time. 

> but the number of subworkflows limits the number of tasks. 

This appears to be the problem.

**Comment by BoPeng** [comment]
The problem with 

```
(multiple chains of ) subworkflow -> step -> (single substep) -> task
```
compared to
```
(single) step -> (multiple) substeps -> tasks
```
is that steps needs to be alive and wait for the tasks to be completed, forcing `subworkflows` to be alive and waiting. This was why we saw "too many open files".

Unless I can really figure out a way to serialize pending steps and subworkflows (e.g. write to disk so that they do not occupy open sockets), this will be a limitation of subworkflows.

**Comment by gaow** [comment]
So for the time being, maybe I can do something smarter (an alternative implementation) to my problem? The step in question is `hybrid_3` in the link below:

https://gaow.github.io/cnv-gene-mapping/workflow/20190717_workflow.html

where that nested workflow calls two languages written in different languages. The problem is that those other steps are used on their own in other workflows, so we cannot *move* their codes to `hybrid_3`. Then any code copied into `hybrid_3` will be redundant.

**Comment by BoPeng** [comment]
Sorry, I will leave it to you to figure out how to adapt to this limitation. 

I think we once had something like no wait mode in that sos does not wait for the tasks to return. The workflow could even continue to execute if the step is not expecting any output, otherwise sos would simply exit, saying all tasks have been submitted. This could be potentially useful but I do not think it worthwhile to bring back such a feature just for this limitation.

**Comment by gaow** [comment]
A comment on this limitation: suppose I have several workflows in a notebook, `a` to `e`, each has independent input and output, and will be submitted to cluster for executing many substeps. Instead of running `sos` command 5 times each command running one of the steps, I would like to write a `default` step like:

```
[default]
sos_run('a')
sos_run('b')
sos_run('c')
sos_run('d')
sos_run('e')
```

But it seems this wouldn't work because if I write like this, then the execution will be limited to the number of CPUs specified in `-j` even if it is meant to actually use `-J` to submit lots of jobs to cluster?

**Comment by BoPeng** [comment]
The limitation is on the number of concurrent subworkflows in a single step, in your case you are running workflows sequentially. If they can be executed concurrently, you should write them as

```
sos_run(['a', 'b', 'c', 'd'])
```
and `4` workflows would be submitted altogether, and this limitation does not apply.
